Bump github/codeql-action from 3.28.3 to 3.28.11 (#276)
rsksmart/devportal: tutorials/tokens/create-a-token.md

> It can make the RSK life easier too because we can configure a RSK network in Truffle.

Reviewer: ```suggestion
Truffle makes it easy to configure a connection to the RSK network.
```

@owans note that this part here doesn't refer to OZ.

[trufflesuite.com/truffle](https://www.trufflesuite.com/truffle)

> # Verify that you can connect to RSK Testnet

Reviewer: Not sure why you're removing this section and moving it below.
I think it is better to keep check for this as early as possible.

Reviewer 2: In my mind, in my logical path I will only do it here...
But I think we need to improve this, wherever it was placed. 
What to do if we can't connect?
We need to provide a solution for this

Reviewer: Well if you can't reach the testnet node like this, there are two possibilities:

1. The node is genuinely down, and there's not much the end user can do but wait until it gets fixed
1. They have a flaky internet connection, or firewall/ proxy that is barring access. For example if their network setup bans connections based on a domains whitelist/blacklist or port number (which is unfortunately quite common in offices). In this case the user should resolve those issues before continuing with the tutorial.

In any case, that's not really the prerogative of their tutorial - the objective here is not to be tech support, but to focus on the materials itself.

```

> The option `--E` is to save dependencies with an exact version rather than using npm's default.

Reviewer: should be either `-E` or  `--save-exact`

Reviewer 2: Yes, you are right!
I know that --E works too, because I got confused with this several times, but that's not what the documentation says.

![package.json](/assets/img/tutorials/create-a-token/image-11.png)

# Verify that you can connect to RSK Testnet

Enter the following command into your terminal. 

If you are using a Windows OS, I suggest to use the Git Bash terminal. 
Download the installer from the [Git official site](https://git-scm.com/).

```shell
curl https://public-node.testnet.rsk.co/1.3.0/ \
  -X POST -H "Content-Type: application/json" \
  --data '{"jsonrpc":"2.0","method":"eth_blockNumber","params":[],"id":1}'
```

This is a very simple query that simply asks what the latest block number is.

You should receive a response similar to the following:

```json
{"jsonrpc":"2.0","id":1,"result":"0xc3f9b"}
```

![eth_blockNumber jsonrpc result](/assets/img/tutorials/create-a-token/image-42.png)

The result is presented in hexadecimal. `0xc3f9b` is the block number `802715`. 
You can consult [explorer.testnet.rsk.co](https://explorer.testnet.rsk.co/) and verify that it is the same result for block number.

![explorer.testnet.rsk.co blockNumber](/assets/img/tutorials/create-a-token/image-43.png)

# Create a mnemonic

> 

Reviewer: The use for BIP39 mnemonics deviates significantly from how it is defined here.
Instead of copying this definition, I would rather simply include a link to the BIP39 standards page.

> 

Reviewer 2: Ok, I will add this link
https://github.com/bitcoin/bips/blob/master/bip-0039.mediawiki

> 

Reviewer: :+1:

  --data '{"jsonrpc":"2.0","method":"eth_blockNumber","params":[],"id":1}'
```

This is a command to query the network for the latest block number.

If all goes well, you'll see an output similar to the following:

```json
{"jsonrpc":"2.0","id":1,"result":"0xc3f9b"}
```

![eth_blockNumber jsonrpc result](/assets/img/tutorials/create-a-token/image-42.png)

The `result` value is presented in hexadecimal. `0xc3f9b` is the block number in hexadecimal, the corresponding decimal is: `802715`. 
To verify the block number, visit [explorer.testnet.rsk.co](https://explorer.testnet.rsk.co/).

![explorer.testnet.rsk.co blockNumber](/assets/img/tutorials/create-a-token/image-43.png)

# Create a mnemonic

> Let's create a mnemonic to generate addresses with it.

Reviewer: ```suggestion
Let's create a mnemonic in order to generate some accounts.
```

> To learn more about it: [Bip39](https://github.com/bitcoin/bips/blob/master/bip-0039.mediawiki)

Reviewer: ```suggestion
To learn more about it: [BIP39](https://github.com/bitcoin/bips/blob/master/bip-0039.mediawiki)
```

> > It is not recommended to be used for any 'real' wallet because it's not a secure way to generate a private key, however we will use this here for learning purposes, and since we're using the Testnet anyway.

Reviewer: Hmm, I don't think this is correct assertion.
BIP39 mnemonics are indeed recommended for usage in a real wallet.
In fact, even hardware wallets support the use of BIP39 mnemonics.

However, what **isn't** recommended for real usage is the way that we are storing them in this tutorial,
being saved to a file on disk in unencrypted form.

> > It is not recommended to be used for any 'real' wallet because it's not a secure way to generate a private key, however we will use this here for learning purposes, and since we're using the Testnet anyway.

Reviewer 2: I think it's not secure generate a private key in a website...

> In the `Generate a random mnemonic` field, select `12 words` and click on the `generate` button.

Reviewer: The link has indeed been included above, but it doesn't have that option, and I think for good reason - you never would want to give out your BIP39 mnemonic to anyone else. (Obviously demos/ workshops like this are an exception.)

## File .secret

> Inside the folder token, create a file named `.secret`.

Reviewer: ```suggestion
Inside the `token` folder, create a file named `.secret`.
```

> Truffle has your own console to run commands.

Reviewer 3: ```suggestion
> You can use the Truffle console to run commands.
```

> ## Verify the connection 

Reviewer: ```suggestion
## Verify the connection
```

> Let’s open a Truffle console to check our connection.

Reviewer 3: ```suggestion
Open a Truffle console to verify the connection.
```

Reviewer 3: Any guide to do this?

Reviewer: > Any guide to do this?

`truffle console --network testnet` - it's in the next instruction

![truffle console](/assets/img/tutorials/create-a-token/image-15.png)

> What we are doing here is telling Truffle to connect to an RSK public testnet node, and giving it control over your recently created account.

Reviewer 3: ```suggestion
This action instructs Truffle to connect to an RSK public testnet node and grants it permission to control the newly created account.
```

## To get our address

> Let’s type this in Truffle console to save in variable `account` the first address generated with our mnemonic:

Reviewer 3: ```suggestion
Insert this command below in Truffle console to save in variable `account` the first address generated with our mnemonic:
```

Reviewer: ```suggestion
Enter the command below into the Truffle console to save the first address generated with our mnemonic into an `account` variable:
```

```javascript
var account = Object.keys(web3.currentProvider.wallets)[0]
```

> The result is your account address. Type this to show it:

Reviewer 3: ```suggestion
The output is your account address. Insert this command to view it:
```

Reviewer: ```suggestion
The output is your account address. Enter this command to view it:
```

```javascript
account
```

> In my example, it is `0x9682725a85f85f097ab368555a286618dc982c99`. Copy this address.

Reviewer 3: ```suggestion
In my example, the output is `0x9682725a85f85f097ab368555a286618dc982c99`. Copy this address.
```

![account address](/assets/img/tutorials/create-a-token/image-16.png)

## Check balance

> Now, let’s check our balance. Run this command in Truffle console:

Reviewer 3: ```suggestion
To check balance, run this command in Truffle console:
```

Reviewer: ```suggestion
To check the balance, run this command in Truffle console:
```

```javascript
(await web3.eth.getBalance(account)).toString()
```

> The balance is 0 and we need tR-BTC. Let's get it in the next step.

Reviewer 3: ```suggestion
The balance is 0, we will purchase tR-BTC in the next step.
```

Reviewer: ```suggestion
The balance is 0 and we need some to pay for gas fees. We shall get some tR-BTC in the next step.
```

![getBalance(account) 0](/assets/img/tutorials/create-a-token/image-44.png)

# TestNet Faucet

> You can get some tR-BTC from 

Reviewer 3: the sentence is incomplete

> # TestNet Faucet
> 
> You can get some tR-BTC from 

Reviewer: ```suggestion
# Testnet Faucet

You can get some tR-BTC from the RSK Testnet faucet.
```

> ## Check balance (again)

Reviewer 3: ```suggestion
## Recheck balance
```

> Let’s check our balance again. Run this command in our Truffle console:

Reviewer 3: ```suggestion
To check balance again, run this command in the Truffle console:
```

```javascript
(await web3.eth.getBalance(account)).toString()
```

![getBalance](/assets/img/tutorials/create-a-token/image-20.png)

> Now I have 50000000000000000, which means that I have 0.05 with 18 decimal places of precision.

Reviewer: > is it 0.05 tR-BTC?

yes

# Deploy a smart contract

First of all, we need to create a a new migrations file where Truffle will find it, containing instructions to deploy the smart contract.

> ## Create file 2_deploy_contracts.js

Reviewer: ```suggestion
## Create token deployment file
```

`My RSK Token` is now published on the RSK Testnet.

> Save the contract address of token, it will be used in the future: 

Reviewer: ```suggestion
Save the contract address of token, it will be used shortly:
```

> Check if our instance is OK.

Reviewer: > Is it truffle instance?

Token instance.

In the code `Token` is the token smart contract, think of it as a class; and `token` is the instance of that token token smart contract, think of it as similar to a singleton pattern (only one instance of a class).

> Put the instance’s name:  `token`, hit `.`, then hit TAB twice to trigger autocomplete. 

Reviewer: ```suggestion
Enter the instance’s name:  `token`, then `.`, without space hit the TAB button twice to trigger auto-complete as seen below. 
```

> This will display the published address, transaction hash of deploy, among other things, including all public variables and methods available.

Reviewer: ```suggestion
This will display the published address of the smart contract, and the transaction hash for its deployment, among other things, including all public variables and methods available.
```

> I hope contract is the right word

Yes, that's right.

## Check the total supply

> Call the function `totalSupply` to check if we have tokens already minted:

Reviewer: ```suggestion
To check if we have tokens already minted, call the `totalSupply` function:
```

## Check the token balance

> Call the function `balanceOf` to check the balance of an account, for example, from account 0:

Reviewer: ```suggestion
To check the balance of an account, call the `balanceOf` function. For example, to check the balance of account 0:
```

It sent a transaction to mint 100 tokens for account 0. 

![token.mint account 0](/assets/img/tutorials/create-a-token/image-45.png)

It is possible to verify the transaction at explorer:

> [0x2162617b34ffcd55cf719cb998e69a33cf115c5d4d58b7ee639c1060fae81355](https://explorer.testnet.rsk.co/tx/0x2162617b34ffcd55cf719cb998e69a33cf115c5d4d58b7ee639c1060fae81355)

Reviewer: ```suggestion
[`0x2162617b34ffcd55cf719cb998e69a33cf115c5d4d58b7ee639c1060fae81355`](https://explorer.testnet.rsk.co/tx/0x2162617b34ffcd55cf719cb998e69a33cf115c5d4d58b7ee639c1060fae81355)
```

The transaction:

> [0x1534230dea0ba07b876dd0ad22fdcb693359de42cb12e5af5e55e17543828a85](https://explorer.testnet.rsk.co/tx/0x1534230dea0ba07b876dd0ad22fdcb693359de42cb12e5af5e55e17543828a85)

Reviewer: ```suggestion
[`0x1534230dea0ba07b876dd0ad22fdcb693359de42cb12e5af5e55e17543828a85`](https://explorer.testnet.rsk.co/tx/0x1534230dea0ba07b876dd0ad22fdcb693359de42cb12e5af5e55e17543828a85)
```

> Also you can get the balance of a specific address, for example,  `0xa52515946DAABe072f446Cc014a4eaA93fb9Fd79`:

Reviewer: ```suggestion
Also you can get the balance of a specific address, for example, `0xa52515946DAABe072f446Cc014a4eaA93fb9Fd79`:
```
